src/models/ctx/ctx.rs

```
@@ -229,66 +264,79 @@ fn authenticate<E: Env + 'static>(auth_request: &AuthRequest) -> Effect {
+                    })
+                    .map(|CollectionResponse { addons, .. }| addons)
+            };
 
-                future::try_join(addon_collection_fut, datastore_library_fut)
-                    .map_ok(move |(addons, library_items)| (auth, addons, library_items))
-            })
-            .map(enclose!((auth_request) move |result| {
-                let internal_msg = Msg::Internal(Internal::CtxAuthResult(auth_request, result));
+            let datastore_library_fut = async {
+                let request = DatastoreRequest {
+                    auth_key: auth.key.to_owned(),
+                    collection: LIBRARY_COLLECTION_NAME.to_owned(),
+                    command: DatastoreCommand::Get {
+                        ids: vec![],
+                        all: true,
+                    },
+                };
 
-                event!(Level::TRACE, internal_message = ?internal_msg);
-                internal_msg
-            }))
-            .boxed_env(),
+                fetch_api::<E, _, _, LibraryItemsResponse>(&request)
+                    .inspect(move |result| {
+                        trace!(?result, ?request, "Get user's Addon Collection request")
+                    })
+                    .await
+                    .map_err(CtxError::from)
+                    .and_then(|result| match result {
+                        APIResult::Ok { result } => Ok(result.0),
+                        APIResult::Err { error } => Err(CtxError::from(error)),
+                    })
+            };
+
+            let (addon_collection_result, datastore_library_result) =
+                future::join(addon_collection_fut, datastore_library_fut).await;
+
+            // lock if the result from fetching the addons has failed
+            let addons_locked = addon_collection_result.is_err();
+            // set the flag to true if fetching of the library has failed
+            let library_missing = datastore_library_result.is_err();
```

i dont like it that errors are swallowed here
if you change the CtxAuthResponse struct as i suggested this could be handled like

  future::try_join(addon_collection_fut, datastore_library_fut)
                    .then(move |(addons_result, library_items_result)| (auth, addons_result, library_items_result))
                    
then fallback to Official_addons somwhere else, not inside the authenticate function

---

Resolved